libs/constrained-generators/src/Constrained/Univ.hs

```diff
@@ -85,18 +101,58 @@ data
   = OneofLeft !(fn fnU as b)
   | OneofRight !(fn' fnU as b)
 
+-- | Build a balanced tree of `Oneof` from a list of function universes.
+-- NOTE: it is important that this is a balanced tree as that reduces the amount of
+-- overhead in `injectFn` and `extractFn` from `O(n)` to `O(log(n))`. Surprisingly,
+-- we've observed this making more than 10% difference in runtime on some generation
+-- tasks even though the list `as` is typically small (< 20 elements).
```

Comment:
So, just reordering a linear list, so that the most common things come earlier in the list could give a 10% improvement, with much simpler code.

Reply:
Not quite, remember that everything in the system is always traversing these Oneof trees to get at the function symbol, with the consequence that its also traversing dictionaries for classes and a whole bunch of other stuff. But crucially it's doing this _for every function_ no matter where it is in the tree.

Yes, if you were to have a big enough benchmark suite you could probably figure out an "optimal" order, but that would in turn be super fragile for the actual user. If they were doing things that happen to appear late in the list, they would get a performance penalty. When you do it like this performance is much more predictable because everything is at the same depth (and that depth is close to what you would have at the start of the linear list).